fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/properties/RecordTypesProperty.java

```diff
@@ -49,73 +47,88 @@
 
     private RecordTypesProperty(@Nonnull PlanContext context) {
         this.context = context;
     }
 
     @Override
-    public boolean visitEnter(@Nonnull PlannerExpression expression) {
+    public boolean shouldVisit(@Nonnull PlannerExpression expression) {
         return expression instanceof RelationalPlannerExpression;
     }
 
     @Override
-    public boolean visitEnter(@Nonnull ExpressionRef<? extends PlannerExpression> ref) {
+    public boolean shouldVisit(@Nonnull ExpressionRef<? extends PlannerExpression> ref) {
         return true;
     }
 
-
+    @Nonnull
     @Override
-    public boolean visitLeave(@Nonnull PlannerExpression expression) {
-        if (expression instanceof RecordQueryScanPlan) {
-            possibleTypeSets.push(context.getMetaData().getRecordTypes().keySet());
-        } else if (expression instanceof RecordQueryIndexPlan) {
-            Index index = context.getIndexByName(((RecordQueryIndexPlan)expression).getIndexName());
-            possibleTypeSets.push(context.getMetaData().recordTypesForIndex(index).stream()
-                    .map(RecordType::getName).collect(Collectors.toSet()));
-        } else if (expression instanceof TypeFilterExpression) {
-            Set<String> childTypeSet = possibleTypeSets.pop();
-            possibleTypeSets.push(Sets.filter(childTypeSet,
-                    ((TypeFilterExpression)expression).getRecordTypes()::contains));
-        } else if (expression instanceof RelationalExpressionWithChildren &&
-                   ((RelationalExpressionWithChildren)expression).getRelationalChildCount() > 1) {
-            // If we have a single child, then there is a reasonable default for how most relational expressions will
-            // change the set of record types (i.e., they won't change them at all). However, if you have several relational
-            // children (like a union or intersection expression) then we must specify some way to combine them.
-            int relationalChildCount = ((RelationalExpressionWithChildren)expression).getRelationalChildCount();
-            if (expression instanceof RecordQueryUnionPlan || expression instanceof RecordQueryIntersectionPlan) {
-                Set<String> union = new HashSet<>();
-                for (int i = 0; i < relationalChildCount; i++) {
-                    union.addAll(possibleTypeSets.pop());
+    public Set<String> evaluateAtExpression(@Nonnull PlannerExpression expression, @Nonnull List<Set<String>> childResults) {
+        // shouldVisit() ensures that we only visit relational planner expressions
+        // If we mess this up, better to find out sooner rather than later.
+        final RelationalPlannerExpression relationalExpression = (RelationalPlannerExpression) expression;
+
+        if (relationalExpression instanceof RecordQueryScanPlan) {
+            return context.getMetaData().getRecordTypes().keySet();
+        } else if (relationalExpression instanceof RecordQueryIndexPlan) {
```

Comment:
Does this to also handle `RecordQueryTextIndexPlan`? Or maybe `RecordQueryPlanWithIndex` or whatever that's called?